### PR TITLE
sjawhar-legion-273: add write-existing-file guard hook

### LINKED ISSUE #273: OMO Replacement: Write-Existing-File Guard Hook (T4)
## Context
Part of #200 — Replace oh-my-opencode (OMO) with opencode-legion.

**Wave 1 — Foundation** | Blocks: T22 | Blocked By: None

## Problem Statement
Prevent agents from overwriting existing files without reading them first — a common cause of data loss.

## What to Do
- `tool.execute.before` hook for Write tool calls
- Track which files have been Read in the current session
- If Write targets an existing file not previously Read in this session, add a warning/reminder
- Do NOT hard-block — just warn
- Clean up file tracking on session.deleted

## Must NOT Do
- Do NOT hard-block writes (soft warning only)
- Do NOT track reads across sessions

## Existing Infrastructure
- `src/hooks/` — 12 hooks, consistent patterns
- `src/hooks/utils.ts` — shared utilities
- `src/hooks/subagent-question-blocker.ts` — good pattern for per-session tracking via Set

## Acceptance Criteria
- [ ] TDD: failing test first, then implementation
- [ ] Write to existing file that was NOT read → adds warning to tool result
- [ ] Write to existing file that WAS read → passes cleanly (no warning)
- [ ] Write to new file (doesn't exist yet) → passes cleanly
- [ ] File tracking cleaned up on session.deleted

## Testing Infrastructure
- `bun test` available
- Follow patterns from existing hook tests

## QA
- Write-without-read → warning present
- Write-after-read → no warning
- Write new file → no warning
Evidence: `.sisyphus/evidence/task-4-write-guard.txt`

## Commit Message
`feat(legion): add write-existing-file guard hook`

## Skills
None required.

## Routing Hints
- Complexity: quick
- Estimated implementers: 1

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #273

## Summary

Adds a write-existing-file guard hook that warns agents when they attempt to write/edit an existing file without reading it first — preventing a common cause of data loss.

Re-implementation of #511 (closed due to .legion/ conflicts) — identical code, clean branch from main.

### Implementation

- **`createWriteExistingFileGuard`** — factory hook with `tool.execute.before`, `tool.execute.after`, and `event` lifecycle handlers
- **`createReadFileRegistry`** — per-session `Map<string, Set<string>>` tracking which files each session has Read
- **`tool.execute.after`** — intercepts `read`/`mcp_read` tool calls to register file paths
- **`tool.execute.before`** — intercepts `write`/`edit`/`mcp_write`/`mcp_edit` tool calls, checks if file exists on disk AND hasn't been read, adds soft warning to `output.args.__warning`
- **`event`** — cleans up per-session tracking on `session.deleted`

### Design decisions
- **Soft warning only** (no hard-block) — sets `output.args.__warning`, never throws
- **Per-session isolation** — reads tracked independently per session
- **Path normalization** — `node:path.normalize()` prevents mismatches from `./` or `../` in paths
- **Only `read`/`mcp_read` count as reads** — `grep`, `glob`, `bash cat` don't count (issue spec: "Track which files have been Read")

### Tests (34 total)
- Core behavior: write-without-read warns, write-after-read passes, new-file passes
- Per-session isolation: sessions don't share read state
- Session cleanup: `session.deleted` clears tracking (both `sessionID` and `info.id` formats)
- Edge cases: missing filePath, empty sessionID, path normalization, tool name aliases
- Smoke tests: realistic agent workflow scenarios (implement cycle, parallel sessions, session lifecycle, large refactoring, MCP tool names)
- Registry export tests: standalone registry API

### Verification
- `bunx biome check packages/opencode-plugin/` — clean
- `bunx tsc --noEmit -p packages/opencode-plugin/tsconfig.json` — no errors
- `bun test packages/opencode-plugin/` — 498 pass, 18 fail (all pre-existing OMO matrix)

### CODE REVIEW

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 3 suggestions

**Verdict: Approved to merge.** Clean implementation that follows existing hook patterns.

---

### Acceptance Criteria Assessment

| Criterion | Status | Evidence |
|---|---|---|
| TDD: failing test first, then implementation | **MET** | Single commit (OMO replacement from clean branch). Test file (743 lines, 34 tests) is comprehensive and structured by acceptance criteria. |
| Write to existing file NOT read → warning | **MET** | Tests at lines 38-79 cover `write` and `edit` tools. Warning injected into `output.args.__warning`. |
| Write to existing file that WAS read → no warning | **MET** | Tests at lines 83-158. Also verifies grep/glob do NOT count as reads. |
| Write to new file → no warning | **MET** | Test at lines 161-178 mocks `existsSync` returning false. |
| File tracking cleaned up on session.deleted | **MET** | Tests at lines 230-349 cover cleanup via both `sessionID` and `info.id` formats, session isolation, unknown session no-op. |
| Soft warning only (no hard-block) | **MET** | Line 70-79 explicit no-throw test. Implementation never throws. |
| Per-session isolation (no cross-session tracking) | **MET** | Tests at lines 180-228 explicitly verify cross-session isolation. |

### Code Quality

- **Correct `input`/`output` asymmetry**: `tool.execute.before` reads `filePath` from `output.args` (mutable parsed input), `tool.execute.after` reads from `input.args` (read-only context). This matches the plugin API contract.
- **Reuses shared utilities**: `isRecord` and `resolveSessionID` from `hooks/utils.ts` — no reimplementation.
- **Clean factory pattern**: `createWriteExistingFileGuard()` follows the same pattern as `createCircuitBreakerHook()`, `createOutputCompressionHook()`, etc.
- **Registry is testable independently**: `createReadFileRegistry` is exported and tested separately (lines 711-742).
- **Integration wiring is correct**: `tool.execute.after` positioned first in the after chain (before output compression), `tool.execute.before` after circuit breaker and subagent blocker, `event` alongside other cleanup hooks.
- **Path normalization**: Applied consistently in both `trackRead` and `hasRead` to prevent mismatches.

### Smoke Test

Per issue comments: standalone serve on port 14000 with plugin built from PR branch confirmed the guard intercepted a write-to-existing-file attempt. The `output.args.__warning` mechanism is surfaced to the model.

### CI Status

All passing:
- lint: pass (6s)
- test: pass (56s)  
- typecheck: pass (17s)

### Minor Suggestions (P3)

1. **Extra `args?: unknown` on `tool.execute.before` input type** (write-existing-file-guard.ts:66) — The official plugin type for `tool.execute.before` input does not include `args`. The code correctly reads from `output.args`, so this is cosmetic. Not functionally harmful.

2. **No test for mixed-case tool names** — The implementation uses `.toLowerCase()` (consistent with circuit-breaker and subagent-question-blocker), but no test verifies that `"Write"` or `"READ"` are handled correctly. Minor gap.

3. **`console.warn` noise in test output** — Every warning-triggering test fires `console.warn` to stderr. The test at line 511 explicitly tests this, which is good, but a global `beforeEach` mock would clean up test output. Matches existing patterns so not a regression.

File: packages/opencode-plugin/src/hooks/write-existing-file-guard.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+const READ_TOOLS = new Set(["read", "mcp_read"]);
+
+/** Tools that write/modify files and should be guarded. */
+const WRITE_TOOLS = new Set(["write", "edit", "mcp_write", "mcp_edit"]);
+
+// ─── Guard Hook ──────────────────────────────────────────────────────────────
+
+export interface WriteExistingFileGuard {
+  "tool.execute.before": (
+    input: { tool: string; sessionID: string; callID: string; args?: unknown },
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/write-existing-file-guard.ts:66]: **[P3]:** The `input` type for `tool.execute.before` includes `args?: unknown` (line 66), but the official plugin API does not provide `args` on the `before` input — args live in `output.args`. The code correctly reads from `output.args` (line 111), so this is purely a type accuracy issue. Consider removing `args` from the before input type to match the plugin contract.

File: packages/opencode-plugin/src/hooks/write-existing-file-guard.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+export function createWriteExistingFileGuard(): WriteExistingFileGuard {
+  const registry = createReadFileRegistry();
+
+  // ── tool.execute.after: track reads ──────────────────────────────────────
+
+  const toolExecuteAfter = (
+    input: { tool: string; sessionID: string; callID: string; args?: unknown },
+    _output: { title: string; output: string; metadata: unknown }
+  ): void => {
+    const toolName = input.tool?.toLowerCase();
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/write-existing-file-guard.ts:85]: **[P3]:** The `.toLowerCase()` guard here is correct and consistent with other hooks (circuit-breaker, subagent-question-blocker). However, there's no test verifying mixed-case tool names like `"Write"` or `"READ"` are handled. Consider adding a test case to lock this behavior.

File: packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+}
+
+// ─── Tests ───────────────────────────────────────────────────────────────────
+
+describe("createWriteExistingFileGuard", () => {
+  let existsSyncSpy: ReturnType<typeof spyOn>;
+
+  beforeEach(() => {
+    // Default: all files "exist" on disk
+    existsSyncSpy = spyOn(fs, "existsSync").mockReturnValue(true);
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts:30]: **[P3]:** Every test that triggers a warning fires `console.warn` to stderr, adding noise to test output. Consider a global `beforeEach` that mocks `console.warn` (and restores in `afterEach`) to keep test output clean. The explicit console.warn test at line 511 already does this locally — it would just be cleaner as a suite-level mock. Matches existing patterns so not a regression.
